spec/unit/coinbase/smart_contract_spec.rb

```diff
@@ -198,6 +198,388 @@
     end
   end
 
+  describe '.read' do
+    subject(:result) do
+      described_class.read(
+        network: network,
+        contract_address: contract_address,
+        method: method_name,
+        abi: abi,
+        args: args
+      )
+    end
+
+    let(:contract_address) { '0x1234567890123456789012345678901234567890' }
+    let(:method_name) { 'testMethod' }
+    let(:args) { {} }
+    let(:abi) { nil }
+
+    before do
+      allow(smart_contracts_api).to receive(:read_contract)
+    end
+
+    describe 'network parameter' do
+      it 'uses the correct normalized network ID' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with('base-sepolia', anything, anything)
+      end
+    end
+
+    describe 'address parameter' do
+      it 'uses the provided contract address' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with(anything, contract_address, anything)
+      end
+    end
+
+    describe 'method parameter' do
+      it 'includes the method name in the request params' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with(anything, anything,
+                                                                          hash_including(method: method_name))
+      end
+    end
+
+    describe 'abi parameter' do
+      context 'when provided' do
+        let(:abi) { [{ 'name' => 'testMethod', 'inputs' => [], 'outputs' => [] }] }
+
+        it 'sends JSON encoded hash as abi in request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(anything, anything,
+                                                                            hash_including(abi: abi.to_json))
+        end
+      end
+
+      context 'when explicitly set to nil' do
+        let(:abi) { nil }
+
+        it 'omits the abi from request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(
+            anything,
+            anything,
+            hash_including(abi: nil)
+          )
+        end
+      end
+
+      context 'when omitted' do
+        subject(:result) do
+          described_class.read(
+            network: network,
+            contract_address: contract_address,
+            method: method_name
+          )
+        end
+
+        it 'omits the abi from request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(
+            anything,
+            anything,
+            hash_including(abi: nil)
+          )
+        end
+      end
+    end
```

Comment:
```suggestion
    let(:abi) { [{ 'name' => 'testMethod', 'inputs' => [], 'outputs' => [] }] }
    
    it 'calls the read_contract API' do
      result
      
      expect(smart_contracts_api)
        .to have_received(:read_contract)
        .with(
          'base-sepolia',
           contract_address,
           {
           hash_including(method: method_name, abi: abi.to_json, args: args.to_json)
        )
    end

    context 'when the ABI parameter is omitted' do
        ...
    end
    
    context 'when the args parameter is omitted' do
    end
    
    context 'when the args paramter is null' do
    end
    
    // Table-driven tests
    [].each do
      context 'when the return value is' do
          ...
      end
    end
```

This is how I would structure the describe / context blocks!

I'm not sure why the rspec linters are not flagging the describe blocks 🤔

Reply:
yeah this is cleaner, good call!